crates/uv-distribution/src/pyproject_mut.rs

```diff
@@ -0,0 +1,154 @@
+use std::fmt;
+use std::str::FromStr;
+
+use thiserror::Error;
+use toml_edit::{Array, DocumentMut, Item, RawString, TomlError, Value};
+
+use pep508_rs::{Pep508Error, Requirement};
+use pypi_types::{LenientRequirement, VerbatimParsedUrl};
+
+use crate::pyproject::PyProjectToml;
+
+/// Raw and mutable representation of a `pyproject.toml`.
+///
+/// This is useful for operations that require editing an existing `pyproject.toml` while
+/// preserving comments and other structure, such as `uv add` and `uv remove`.
+pub struct PyProjectTomlMut {
+    doc: DocumentMut,
+}
+
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error("Failed to parse `pyproject.toml`")]
+    Parse(#[from] TomlError),
+    #[error("Dependencies in pyproject.toml are malformed")]
+    MalformedDependencies,
+}
+
+impl PyProjectTomlMut {
+    /// Initialize a `PyProjectTomlMut` from a `PyProjectToml`.
+    pub fn from_toml(pyproject: &PyProjectToml) -> Result<Self, Error> {
+        Ok(Self {
+            doc: pyproject.raw.parse()?,
+        })
+    }
+
+    /// Adds a dependency.
+    pub fn add_dependency(&mut self, req: &Requirement) -> Result<(), Error> {
+        let deps = &mut self.doc["project"]["dependencies"];
+        if deps.is_none() {
+            *deps = Item::Value(Value::Array(Array::new()));
+        }
+        let deps = deps.as_array_mut().ok_or(Error::MalformedDependencies)?;
+
+        // Try to find a matching dependency.
+        let mut to_replace = None;
+        for (i, dep) in deps.iter().enumerate() {
+            if let Some(dep) = dep.as_str().and_then(|dep| parse_requirement(dep).ok()) {
+                if dep.name.as_ref().eq_ignore_ascii_case(req.name.as_ref()) {
+                    to_replace = Some(i);
+                    break;
+                }
+            }
+        }
+
+        match to_replace {
+            Some(i) => {
+                deps.replace(i, req.to_string());
+            }
+            None => deps.push(req.to_string()),
+        }
+
+        reformat_array_multiline(deps);
+
+        Ok(())
+    }
+
+    /// Removes a dependency.
+    pub fn remove_dependency(&mut self, req: &Requirement) -> Result<Option<Requirement>, Error> {
+        let deps = &mut self.doc["project"]["dependencies"];
+        if deps.is_none() {
+            return Ok(None);
+        }
+
+        let deps = deps.as_array_mut().ok_or(Error::MalformedDependencies)?;
+
+        // Try to find a matching dependency.
+        let mut to_remove = None;
+        for (i, dep) in deps.iter().enumerate() {
+            if let Some(dep) = dep.as_str().and_then(|dep| parse_requirement(dep).ok()) {
+                if dep.name.as_ref().eq_ignore_ascii_case(req.name.as_ref()) {
```

Comment:
What if a dep is included multiple times? E.g.:

```toml
dependencies = [
  "flask > 2 ; python_version >= '3.6'",
  "flask < 2 ; python_version < '3.6'",
]
```

Reply:
Hmm it would only remove the first occurrence then. I think Rye has this bug as well?

Reply:
Updated to remove all occurrences. Also changed `uv remove` to take package names, not full requirements, because otherwise `uv remove flask>2` becomes more complicated. I think package names make more sense anyways? Unless we want to support removing unnamed requirements by their URL?